src/plugins/detect-inconsistency/detect-inconsistency-utils.ts

```diff
@@ -0,0 +1,198 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  ManuscriptEditorState,
+  ManuscriptNode,
+  nodeNames,
+  schema,
+} from '@manuscripts/transform'
+import { NodeSelection } from 'prosemirror-state'
+import { Decoration, DecorationSet } from 'prosemirror-view'
+
+import { getBibliographyPluginState } from '../bibliography'
+import { footnotesKey } from '../footnotes'
+import { objectsKey } from '../objects'
+
+export type Warning = {
+  type: 'warning'
+  category: 'missing-reference' | 'empty-content'
+  severity: 'error' | 'warning'
+  message: string
+  node: ManuscriptNode
+  pos: number
+}
+
+export type PluginState = {
+  decorations: DecorationSet
+  warnings: Array<Warning>
+  showDecorations: boolean
+}
+
+export const createDecoration = (
+  node: ManuscriptNode,
+  pos: number,
+  selectedPos: number | null
+) => {
+  const classNames = ['warning-highlight']
+  if (selectedPos === pos) {
+    classNames.push('selected-suggestion')
+  }
+  return Decoration.node(pos, pos + node.nodeSize, {
+    class: classNames.join(' '),
+    'data-inconsistency-type': 'warning',
+  })
+}
+
+export const getNodeDescription = (node: ManuscriptNode): string => {
+  return nodeNames.get(node.type) || node.type?.name || ''
+}
+
+export const addWarning = (
+  warnings: Warning[],
+  decorations: Decoration[],
+  node: ManuscriptNode,
+  pos: number,
+  category: Warning['category'],
+  severity: Warning['severity'],
+  message: string,
+  showDecorations: boolean,
+  selectedPos: number | null
+) => {
+  warnings.push({
+    type: 'warning',
+    category,
+    severity,
+    message,
+    node,
+    pos,
+  })
+  showDecorations && decorations.push(createDecoration(node, pos, selectedPos))
+}
+
+export const buildPluginState = (
+  state: ManuscriptEditorState,
+  showDecorations: boolean
+): PluginState => {
+  const warnings: Warning[] = []
+  const decorations: Decoration[] = []
+
+  try {
+    const selection = state.selection
+    let selectedPos: number | null = null
+
+    if (selection instanceof NodeSelection) {
+      selectedPos = selection.from
+    }
+
+    const pluginStates = {
+      bibliography: getBibliographyPluginState(state)?.bibliographyItems,
+      objects: objectsKey.getState(state),
+      footnotes: footnotesKey.getState(state),
+    }
+
+    state.doc.descendants((node, pos) => {
+      if (node.type === schema.nodes.title) {
+        const isEmpty = node.content.size === 0
+        if (isEmpty) {
+          addWarning(
+            warnings,
+            decorations,
+            node,
+            pos,
+            'empty-content',
+            'warning',
+            `${getNodeDescription(node)} is empty`,
```

Comment:
is it worth passing it as an argument, looks like addWarning can get this info itself by calling getNodeDescription, since it gets 'node' passed anyway. It's a lot of arguments that we pass and it would be nice to optimise it a bit.